projects/hathor-wallet-headless/0004-tx-transaction.md

```diff
@@ -0,0 +1,493 @@
+	- [NOTE-2]: It may miss the index
+
+>[!NOTE]
+>Decision: A miss here means an input will not be added to the list of inputs. It can drive the final transaction to be invalid, but the template itself will be generated nevertheless.
+
+#### `TokenOutputTemplateInstruction`
+[tokenoutputtemplateinstruction]: #tokenoutputtemplateinstruction
+
+```ts
+interface TokenOutputTemplateInstruction extends InputOutputTemplateInstruction, BaseTemplateInstruction {
+  amount: number;
+  token?: string;
+  address?: string;
+  authority?: `mint` | `melt`;
+  check_address?: boolean;
+}
+```
+
+- `amount`: amount of tokens
+- `token`: which token to select, defaults to native token, in practice `00` (HTR)
+- `address`: create an output script for the address, if not present will get one from the wallet
+	- [NOTE-1]: It will get a not used address from the wallet
+- `authority`: `mint` or `melt`, if present will create the desired authority output
+	- [NOTE-2]: If the required input is not present, then it will generate an invalid transaction
+- `check_address`: whether to check that the address is from the wallet, defaults to `false`
+	- [NOTE-3]: If `true` and address is not from the wallet, then the interpreter should fail short
+
+>[!NOTE]
+>Decision Call: Should we keep the default behavior or `address` or make it more strict? Like, if there isn't any `address` value, then fail short.
+>
+>To counter the behavior of a stricter `address` we can add a flag to let the user choose the desired behavior, either get a new address or get an address by index.
+>
+>Or we can make `address` a poli typed property which can accept `number` for indexes, the string `new` or a general `string` to denote an address.
+
+#### `DataOutputTemplateInstruction`
+
+```ts
+interface DataOutputTemplateInstruction extends InputOutputTemplateInstruction, BaseTemplateInstruction {
+  data: string;
+}
+```
+
+- `data`: UTF-8 encoded string of data
+#### `RawOutputTemplateInstruction`
+
+```ts
+interface RawOutputTemplateInstruction extends InputOutputTemplateInstruction, BaseTemplateInstruction {
+  amount: number;
+  script: string;
+  token?: string;
+  authority?: `mint` | `melt`;
+}
+```
+
+- `amount`: amount of tokens
+- `script`: base64 encoded script
+- `token`: which token to select, defaults to native token, in practice `00` (HTR)
+- `authority`: `mint` or `melt`, if present will create the desired authority output
+	- [NOTE-1]: If the required input is not present, then it will generate an invalid transaction
+
+#### `ShuffleActionTemplateInstruction`
+[shuffleactiontemplateinstruction]: #shuffleactiontemplateinstruction
+
+```ts
+interface RawOutputTemplateInstruction extends ActionTemplateInstruction, BaseTemplateInstruction {
+}
+```
+
+It will apply an array shuffle action over the selected target.
+
+#### `FillChangeActionTemplateInstruction`
+
+```ts
+interface RawOutputTemplateInstruction extends ActionTemplateInstruction, BaseTemplateInstruction {
+}
+```
+
```

Comment:
I believe this action should have an optional field `address`, which is the change address for the outputs created. What do you think?

Reply:
Sure! Nice catch. Thanks!

Reply:
Will add `changeAddress` on `UtxoSelectInstruction`

Reply:
Done